name-registry/src/main.sw

```diff
@@ -0,0 +1,154 @@
+contract;
+
+dep data_structures;
+dep errors;
+dep events;
+dep interface;
+
+use data_structures::Record;
+use errors::{AssetErrors, AuthorisationErrors, ValidityErrors};
+use events::{
+    IdentityChangedEvent,
+    NameRegisteredEvent,
+    OwnerChangedEvent,
+    RegistrationExtendedEvent,
+};
+use interface::NameRegistry;
+use std::{
+    block::timestamp,
+    chain::auth::msg_sender,
+    constants::BASE_ASSET_ID,
+    context::call_frames::msg_asset_id,
+    context::msg_amount,
+    logging::log,
+    storage::StorageMap,
+};
+
+storage {
+    /// A mapping of names to an option of records, with a none representing an unregistered name
+    names: StorageMap<str[8], Option<Record>> = StorageMap {},
+}
+
+// TODO: Change the static 8 length str with a dynamic string when possible
+impl NameRegistry for Contract {
+    #[storage(read, write)]
+    fn extend(name: str[8], duration: u64) {
+        require(storage.names.get(name).is_some(), ValidityErrors::NameNotRegistered);
+        require((duration / 100) * price_per_hundred <= msg_amount(), AssetErrors::InsufficientPayment);
+        require(msg_asset_id() == BASE_ASSET_ID, AssetErrors::IncorrectAssetSent);
+
+        let previous_record = storage.names.get(name).unwrap();
```

Comment:
I think you should be able to use struct destructuring like `let Record { expiry, identity, owner } = storage...` and then reference the variables directly below without having to do `previous_record.`